planner/core/plan_cacheable_checker.go

```diff
@@ -217,6 +217,11 @@ var nonPrepCacheCheckerPool = &sync.Pool{New: func() any { return &nonPreparedPl
 
 // NonPreparedPlanCacheableWithCtx checks whether this SQL is cacheable for non-prepared plan cache.
 func NonPreparedPlanCacheableWithCtx(sctx sessionctx.Context, node ast.Node, is infoschema.InfoSchema) (ok bool, reason string) {
+	_, isSelect := node.(*ast.SelectStmt)
```

Comment:
The `isSelect`  is true for the `select for update` statement which should be also rejected.

Reply:
> Maybe https://github.com/pingcap/tidb/blob/master/parser/ast/util.go#L24 could be used to check if it's read only.

I'm afraid we can't... non-prep cache is very performance sensitive, and this function `IsReadOnly` let the optimizer have to iterate the whole AST one more time, which can cause some extra overhead.